test/extensions/filters/http/ext_proc/filter_test.cc

```diff
@@ -2261,6 +2272,198 @@ TEST_F(HttpFilterTest, FailOnInvalidHeaderMutations) {
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Path()->value().getStringView(), "/mutated_path/bluh");
+        EXPECT_EQ(headers.get(Envoy::Http::LowerCaseString("foo"))[0]->value().getStringView(),
+                  "gift-from-external-server");
+        mock_filter->callbacks_->sendLocalReply(Http::Code::OK, "Direct response from mock filter.",
+                                                nullptr, absl::nullopt, "");
+        return Http::FilterHeadersStatus::StopIteration;
+      }));
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, _))
+      .WillOnce(Invoke([&](const Http::ResponseHeaderMap& headers, bool end_stream) -> void {
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Status()->value().getStringView(), "200");
+      }));
+  EXPECT_CALL(response_encoder_, encodeData(_, _))
+      .WillOnce(Invoke([&](Buffer::Instance& data, bool end_stream) -> void {
+        EXPECT_TRUE(end_stream);
+        EXPECT_EQ(data.toString(), "Direct response from mock filter.");
+      }));
+  // Start the request.
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> Http::Status {
+    EXPECT_EQ(data.length(), 5);
+    data.drain(5);
+
+    HttpConnectionManagerImplMixin::decoder_ = &conn_manager_->newStream(response_encoder_);
+    Http::RequestHeaderMapPtr headers{new Http::TestRequestHeaderMapImpl{
+        {":authority", "host"}, {":path", "/bluh"}, {":method", "GET"}}};
+    HttpConnectionManagerImplMixin::decoder_->decodeHeaders(std::move(headers), false);
+    Buffer::OwnedImpl request_body("Definitely more than 10 bytes data.");
+    HttpConnectionManagerImplMixin::decoder_->decodeData(request_body, true);
+    // Now external server returns the request header response.
+    auto response = std::make_unique<ProcessingResponse>();
+    auto* headers_response = response->mutable_request_headers();
+    auto* hdr = headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+    hdr->mutable_header()->set_key("foo");
+    hdr->mutable_header()->set_value("gift-from-external-server");
+    hdr = headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+    hdr->mutable_header()->set_key(":path");
+    hdr->mutable_header()->set_value("/mutated_path/bluh");
+    HttpFilterTest::stream_callbacks_->onReceiveMessage(std::move(response));
+
+    return Http::okStatus();
+  }));
+
+  Buffer::OwnedImpl fake_input("hello");
+  conn_manager_->onData(fake_input, false);
+}
+
+// Test proves that when encodeData(data, end_stream=true) is called before headers response is
+// returned, ext_proc filter will buffer the data in the ActiveStream buffer without triggering a
+// buffer over high watermark call, which ends in a 500 error on response path.
+TEST_F(HttpFilter2Test, LastEncodeDataCallExceedsStreamBufferLimitWouldJustRaiseHighWatermark) {
+  initialize(R"EOF(
```

Comment:
This `initialize()` call is different than the one in `LastDecodeDataCallExceedsStreamBufferLimitWouldJustRaiseHighWatermark`--is that because this test relies on `request_header_mode: "SKIP"` to simplify the request path, which isn't the focus of this test? If so, suggest adding a comment to this effect, to help a future reader who doesn't have context.

Reply:
right, this test is to test against the "response path" behavior. 
The "external server" in this test only handles "response headers". Comment added.

Reply:
Ok, SG